ceresdb/route.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package ceresdb
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/CeresDB/ceresdb-client-go/types"
+	"github.com/CeresDB/ceresdb-client-go/utils"
+)
+
+type routeClient struct {
+	opts       options
+	endpoint   string
+	rpcClient  *rpcClient
+	routeCache sync.Map // metric -> *Route
+	gcTimes    int64
+}
+
+type routeWrapper struct {
+	route   types.Route
+	lastHit int64
+}
+
+func newRouteClient(endpoint string, rpcClient *rpcClient, opts options) *routeClient {
+	rc := &routeClient{
+		opts:       opts,
+		endpoint:   endpoint,
+		rpcClient:  rpcClient,
+		routeCache: sync.Map{},
+	}
+
+	utils.GoTickerWithRecover(rc.gc, time.Second*time.Duration(opts.RouteGcPeriodSecond), context.Background(),
+		true, opts.Logger)
+
+	return rc
+}
+
+func (c *routeClient) RouteFor(metrics []string) (map[string]types.Route, error) {
+	if len(metrics) == 0 {
+		return nil, errors.New("Null route metrics")
+	}
+
+	local := make(map[string]types.Route, len(metrics))
+	misses := make([]string, 0, len(metrics))
+
+	for _, metric := range metrics {
+		if route, ok := c.loadRoute(metric); ok {
+			local[metric] = route
+		} else {
+			misses = append(misses, metric)
+		}
+	}
+
+	if len(misses) == 0 {
+		return local, nil
+	}
+
+	if err := c.RouteFreshFor(misses); err != nil {
+		return nil, err
+	}
+
+	for _, metric := range misses {
+		if route, ok := c.loadRoute(metric); ok {
+			local[metric] = route
+		} else {
+			return nil, fmt.Errorf("Route not found for metric %s", metric)
+		}
+	}
+	return local, nil
+}
+
+func (c *routeClient) RouteFreshFor(metrics []string) error {
+	routes, err := c.rpcClient.Route(c.endpoint, metrics)
+	if err != nil {
+		return err
+	}
+
+	for _, route := range routes {
+		c.storeRoute(route.Metric, route)
+	}
+	return nil
+}
+
+func (c *routeClient) ClearRouteFor(metrics []string) {
+	if c.opts.LoggerDebug {
+		c.opts.Logger.Write([]byte(fmt.Sprintf("[%d] Clean metrics %v route for refresh code\n", c.gcTimes, metrics)))
+	}
+	for _, metric := range metrics {
+		c.routeCache.Delete(metric)
+	}
+}
+
+// There is no need to lock the gc, although this does not guarantee that the objects being cleaned are always the oldest,
+// nor that the number being cleaned is accurate; But this is enough for a cache, on the contrary, locking here will affect
+// the read and write performance
+func (c *routeClient) gc() {
```

Comment:
How about use a lru library to do this job?

It will make our code less complex.

Reply:
Adding lru can indeed make the code simpler, but at the same time do more things when sdk query. I think in this scenario, accurate lru is not needed, a non-fine cleanup rule is enough

Reply:
>  but at the same time do more things when sdk query.

Perf is not what we aim for at the moment.

> a non-fine cleanup rule is enough
This gc method seems complex for me, and there are no tests to guarantee its correctness

Reply:
route_test is done for this case

Reply:
> How about use a lru library to do this job?
> 
> It will make our code less complex.

I aggree with @jiacai2050 ,  the route module is not a performance bottleneck, it is better to keep the code simple.

Reply:
use github.com/hashicorp/golang-lru instead of this job